### PR TITLE
Move around librpm calls, add `setreuid()` safety net

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## Proposed commit message

This is an odd one. It turns out that the BDB backend used by older `rpm` installs is...really easy to corrupt, and also prone to a variety of crashes should you get too aggressive with spinning up multiple threads to access the RPM backend. The only way I've found to actually solve this in auditbeat is to use `setreuid()` to drop out of root before we make any librpm calls; this prevents librpm from touching any lock files or opening any database files in write mode, which means we can't corrupt the DB if auditbeat stops/crashes, and it also seems to reduce the odds of certain crashes originating from within the guts of librpm/libdb, as we can't perform any of the integrity checking/cleanup operations that we would if the database state files were writable. 

The resulting `package.rpm_drop_to_uid` is not enabled by default, as we need a UID to drop to, and also setting one thread to a different user is (according to the man page for `setuid`) not POSIX-compliant. This is meant as a safety net for users on older linux distros that haven't migrated to sqlite for the RPM backend.

This also moves around some of the librpm calls, we now call `rpmReadConfigFiles()` earlier in the startup process, and we also call `rpmtsSetRootDir()`. This mostly prevents a few confusing error messages when librpm fails, but I woudn't be surprised if it was causing some more subtle bugs as well.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact

This will not alter any default behavior. 

## How to test

You can test this on any linux distro to confirm that no existing behavior has changed. 

However, to test what this is actually meant to prevent, you'll need to spin up a Linux box that uses rpm/BDB. This should be RHEL8 or any RHEL8-derivatives (Rocky8, etc). Then:

1)  Set run `go test -c` in the `package` directory
2) run multiple instances of the `TestWithSuid` test as root: `sudo ./package.test -test.v -test.run TestWithSuid` at once: `for i in {1..30}; do ./package.test -test.v -test.run TestWithSuid & done`
3) Make sure nothing crashes
4) If librpm segfaults anyway, make sure the DB isnt't corrupted: `sudo rpm -qa`.

### CODE REVIEW

**Comment by fearful-symmetry** [comment]
So, uh, the linter environment doesn't seem to have the rpm build headers, which is...annoying.

File: x-pack/auditbeat/module/system/package/package.go [hunk 1]
```diff
@@ -481,26 +490,71 @@ func storePackages(bucket datastore.Bucket, packages []*Package) error {
...
+					ms.log.Debugf("setreuid has failed; package query thread will remain locked")
+				}
+			}()
 
-		rpmPackages, err := listRPMPackages()
-		if err != nil {
-			return nil, fmt.Errorf("error getting RPM packages: %w", err)
+			minus1 := -1
+			currentUID := os.Getuid()
+			_, _, serr := syscall.Syscall(syscall.SYS_SETREUID, uintptr(minus1), uintptr(*ms.config.PackageSuidDrop), uintptr(minus1))
```
[the remark this answers is not in the document]
Comment (nicholasberlin) [reply x-pack/auditbeat/module/system/package/package.go:527]: Oh, it seems the go interface forces it.

File: x-pack/auditbeat/module/system/package/package_test.go [hunk 2]
```diff
@@ -12,21 +12,49 @@ import (
...
+
+	count := 20
+	waiter := sync.WaitGroup{}
+	waiter.Add(count)
+	t.Logf("Starting...")
+	for i := 0; i < count; i++ {
+		inner := i
+		go func() {
+			defer waiter.Done()
+			testMs := MetricSet{
```
Comment (nicholasberlin) [inline x-pack/auditbeat/module/system/package/package_test.go:44]: Why not set the new config option here?

File: x-pack/auditbeat/module/system/package/rpm_others.go [hunk 3]
```diff
@@ -8,7 +8,7 @@ package pkg
 
 import "errors"
 
-func listRPMPackages() ([]*Package, error) {
+func listRPMPackages(_ bool) ([]*Package, error) {
```
Comment (nicholasberlin) [inline x-pack/auditbeat/module/system/package/rpm_others.go:11]: Please forgive my golang naivete, but what is this change?

Comment (nicholasberlin) [reply x-pack/auditbeat/module/system/package/rpm_others.go:11]: Oh, I see. This is an implementation that ignores the parameter.

**Review by nicholasberlin — APPROVED** [review]
A few minor things, but LGTM if CI passes.

File: x-pack/auditbeat/module/system/package/package.go [hunk 4]
```diff
@@ -233,6 +235,13 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 		ms.log.Debug("No state timestamp found.")
 	}
 
+	if config.PackageSuidDrop != nil {
+		if os.Getuid() != 0 && int(*config.PackageSuidDrop) != os.Getuid() {
+			return nil, fmt.Errorf("package.rpm_drop_to_suid is set to %d, but we're running as a different non-root user", config.PackageSuidDrop)
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply x-pack/auditbeat/module/system/package/package.go:240]: `config.PackageSuidDrop` is a pointer, we're dereferencing to an int.

File: x-pack/auditbeat/module/system/package/package.go [hunk 5]
```diff
@@ -233,6 +235,13 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
...
+		}
+		ms.log.Debugf("Dropping to EUID %d for RPM API calls", *ms.config.PackageSuidDrop)
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply x-pack/auditbeat/module/system/package/package.go:242]: Ah, it doesn't matter, but yeah, should keep it consistent.

File: x-pack/auditbeat/module/system/package/package.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicholasberlin) [reply x-pack/auditbeat/module/system/package/package.go:240]: Oh, I mean for the log message.

**Review by dliappis — APPROVED** [review]
LGTM from eng prod PoV, thank you

**Pull request merged**